### PR TITLE
Animated emoji support

### PR DESCRIPTION
**Pull request by a1batross**
Only animates GIF, as it's the only animated format supported by Android's internal drawable decoder.

### CODE REVIEW

**Comment by a1batross** [comment]
Fixes #472

Running this always enabled since October, it doesn't seem to significantly decrease performance.

File: app/src/main/java/com/keylesspalace/tusky/util/CustomEmojiHelper.kt [hunk 1]
```diff
@@ -45,7 +46,9 @@ fun CharSequence.emojify(emojis: List<Emoji>?, view: View) : CharSequence {
     if(emojis.isNullOrEmpty())
         return this
 
-    val builder = SpannableStringBuilder.valueOf(this)
+    val builder = SpannableString.valueOf(this)
```
Comment (connyduck) [inline app/src/main/java/com/keylesspalace/tusky/util/CustomEmojiHelper.kt:49]: Why this change?

File: app/src/main/java/com/keylesspalace/tusky/util/CustomEmojiHelper.kt [hunk 2]
```diff
@@ -45,7 +46,9 @@ fun CharSequence.emojify(emojis: List<Emoji>?, view: View) : CharSequence {
...
+    val pm = PreferenceManager.getDefaultSharedPreferences(view.context)
+    val animate = pm.getBoolean(PrefKeys.ANIMATE_CUSTOM_EMOJIS, false)
```
Comment (connyduck) [inline app/src/main/java/com/keylesspalace/tusky/util/CustomEmojiHelper.kt:51]: The setting should be queried once per Activity and passed into the function as a Boolean param, as we try to do with other settings. `StatusDisplayOptions` is used in most places for this.

File: app/src/main/java/com/keylesspalace/tusky/util/CustomEmojiHelper.kt [hunk 3]
```diff
@@ -16,12 +16,11 @@
...
 
-import android.graphics.Bitmap
 import android.graphics.Canvas
 import android.graphics.Paint
-import android.graphics.drawable.BitmapDrawable
-import android.graphics.drawable.Drawable
-import android.text.SpannableStringBuilder
+import android.graphics.drawable.*
+import android.text.SpannableString
+import android.text.Spanned
```
Comment (connyduck) [inline app/src/main/java/com/keylesspalace/tusky/util/CustomEmojiHelper.kt:23]: unused import

**Review by connyduck — CHANGES_REQUESTED** [review]
😍

File: app/src/main/java/com/keylesspalace/tusky/util/CustomEmojiHelper.kt (hunk 1 above)
Comment (a1batross) [reply app/src/main/java/com/keylesspalace/tusky/util/CustomEmojiHelper.kt:49]: Oh, that seems to be a part of older research when I tried to make this custom emojis to not ANR. 

There is no need in SSB, when we're not modifying the text, so we can use just SpannableString.

Despite this change doesn't seem to fix #1959 and the core issue was text line breaking, I still think using SS rather than SSB is pretty harmless. The only place where SSB must be used is LinkHelper [here](https://github.com/tuskyapp/Tusky/blob/develop/app/src/main/java/com/keylesspalace/tusky/util/LinkHelper.java#L127).

File: app/src/main/java/com/keylesspalace/tusky/util/CustomEmojiHelper.kt (hunk 2 above)
Comment (a1batross) [reply app/src/main/java/com/keylesspalace/tusky/util/CustomEmojiHelper.kt:51]: Correct. Gonna change it then.

**Comment by a1batross** [comment]
Now it's queried only once (in most places), somewhere it's refactored to also do the same with animateAvatars option.

File: app/src/main/java/com/keylesspalace/tusky/util/CustomEmojiHelper.kt (hunk 1 above)
Comment (connyduck) [reply app/src/main/java/com/keylesspalace/tusky/util/CustomEmojiHelper.kt:48]: The reason why SSB was used here is because when the emojify method is used together with the one in LinkHelper, `valueOf` saves one instance creation because an existing one is reused. Not sure if it has a real impact though.

Comment (a1batross) [reply app/src/main/java/com/keylesspalace/tusky/util/CustomEmojiHelper.kt:48]: That makes sense. Gonna revert it then.

Still I think using SSB is quite overhead and even LinkHelper should not use it but that's for another discussion.

**Review by connyduck — APPROVED** [review]
